toolbox/HBTReader.py

```diff
@@ -1,279 +1,686 @@
+
+        if selection is None:
+            selection = np.s_[:]
+        else:
+            trans_index = isinstance(selection, numbers.Integral)
+
+        if type(selection) is list:
+            selection = tuple(selection)
+
+        with self.OpenFile(isnap, filetype="HaloSize") as hostfile:
+            nsub = hostfile["HostHalos"].shape[0]
+            if trans_index:
+                if offset + nsub > selection:
+                    try:
+                        hosthalos.append(
+                            hostfile["HostHalos"][selection - offset])
+                    except ValueError:
+                        pass
+                offset += nsub
+            else:
+                try:
+                    hosthalos.append(hostfile["HostHalos"][selection])
+                except ValueError:
+                    pass
+
+        if len(hosthalos):
+            hosthalos = np.hstack(hosthalos)
+        else:
+            hosthalos = np.array(hosthalos)
+
+        return hosthalos
+
+    def GetHostHalo(self, HostHaloId, isnap=-1):
+        """Returns spatial information of a specific FoF group.
+
+        Arguments:
+            HostHaloId (int): host halo id
+            isnap (int): snapshot number
+
+        Returns:
+            (numpy.ndarray) single FoF group in a structured array
+        """
+
+        return self.LoadHostHalos(isnap, HostHaloId)[0]
+
+    def GetSubsOfHost(self, HostHaloId, isnap=-1):
+        """Loads all subhaloes belonging to a host halo.
+
+        Uses information from ``Membership/GroupedTrackIds``.
+
+        Arguments:
+            HostHaloId (int): row number
+            isnap (int): (default = -1)
+
+        Returns:
+            (numpy.ndarray): array of ``TrackIds``, or empty array if FoF does
+                not host subhalos
+        """
+        with self.OpenFile(isnap) as subfile:
+            trackIds = subfile['Membership/GroupedTrackIds'][HostHaloId]
+
+        try:
+            result = np.hstack(
+                [self.GetSub(trackId, isnap) for trackId in trackIds])
+        except ValueError:
+            result = np.array([])
+
+        return result
+
+    def GetMergerTree(self, HostHaloId, isnap=-1, file=None):
+        """Builds a FoF merger tree starting at a host halo ID.
+
+        Prints a graph which can be consumed by [Dot](graphviz.org] software.
+
+        Arguments:
+            HostHaloId (int): a halo at which to root a tree
+            isnap (int): (default = -1) snapshot number
+            file (File): (default=None) if not ``None``, Dot diagram will be generated
+                on the fly
+
+        Returns:
+            (list): multiply embedded list with a tree, i.e. ``[1, [2, 3]]``
+        """
+
+        progenitors = self.GetHostProgenitors(HostHaloId, isnap, file)
+
+        loggging.debug(
+            "Halo %d at %d with %d progenitor(s)" %
+            (HostHaloId, isnap, progenitors.size))
+
+        # TODO: print TrackId as graph edge label
```

Comment:
better print to stdout if no file:
if file is None:
 file=sys.stdout

Reply:
Done, changed the default `file` to `sys.stdout`.